jwst/pipeline/calwebb_spec2.py

```diff
@@ -336,3 +284,115 @@ def process_exposure_product(
         )
 
         return result
+
+    def _step_verification(self, exp_type, members_by_type, multi_int):
+        """Verify whether requested steps can operated on the given data
+
+        Thought ideally this would all be controlled through the pipeline
+        parameters, the desire to keep the number of config files down has
+        pushed the logic into code.
+
+        Once step and pipeline parameters are retrieved from CRDS, this
+        logic can be removed.
+        """
+
+        # Check for image-to-image background subtraction can be done.
+        if not self.bkg_subtract.skip:
+            if exp_type in WFSS_TYPES or len(members_by_type['background']) > 0:
+
+                if exp_type in WFSS_TYPES:
+                    members_by_type['background'] = []           # will be overwritten by the step
+
+                # Setup for saving
+                self.bkg_subtract.suffix = 'bsub'
+                if multi_int:
+                    self.bkg_subtract.suffix = 'bsubints'
+
+                # Backwards compatibility
+                if self.save_bsub:
+                    self.bkg_subtract.save_results = True
+            else:
+                self.log.debug('Science does not allow direct background subtraction. Skipping "bkg_subtract".')
+                self.bkg_subtract.skip = True
+
+        # Check for imprint subtraction.
+        imprint = members_by_type['imprint']
+        if not self.imprint_subtract.skip:
+            if exp_type in ['NRS_MSASPEC', 'NRS_IFU'] and \
+               len(imprint) > 0:
+                if len(imprint) > 1:
+                    self.log.warning('Wrong number of imprint members')
+                members_by_type['imprint'] = imprint[0]
+            else:
+                self.log.debug('Science does not allow imprint processing. Skipping "imprint_subtraction".')
+                self.imprint_subtract.skip = True
+
+        # Check for NIRSpec MSA bad shutter flagging.
+        if not self.msa_flagging.skip and exp_type not in ['NRS_MSASPEC', 'NRS_IFU']:
+            self.log.debug('Science does not allow MSA flagging. Skipping "msa_flagging".')
+            self.msa_flagging.skip = True
```

Comment:
So what happens if someone creates an off-line ASN table that has multiple products defined, just as an easy way to process a bunch of exposures all at once, in a loop, and there's a mixture of exposures that allow for and don't allow for a given step like msa_flagging? Will setting the skip param to `True` for the first exposure in the list then cause it to get skipped for all remaining exposures or will the vetting be started over from scratch for each exposure?

Encountered this problem in the `calwebb_coron3` pipeline with the `outlier_detection` step. Once the `skip` param is set to `True` for the first exposure, it doesn't even look to see if it's possible to apply to subsequent exposures.

Reply:
Yes, the described scenario is exactly what would happen. This is a general problem rooted in the instantiate-once/use-many-times model of Steps. For this particular instance, IMHO, the `spec2` looping over products was a mistake from the start. There are a number of approaches on this, see JP-1580, and it is probably best to deal with separately.